cpp/src/cider/common/base/StringRef.h

```diff
@@ -0,0 +1,279 @@
+
+#include <cassert>
+#include <functional>
+#include <iosfwd>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include <common/base/unaligned.h>
+#include <common/contrib/cityhash102/include/city.h>
+
+#include <emmintrin.h>
+#include <nmmintrin.h>
+#include <smmintrin.h>
+
+namespace cider::hashtable {
+
+/**
+ * The std::string_view-like container to avoid creating strings to find substrings in the
+ * hash table.
+ */
+struct StringRef {
+  const char* data = nullptr;
+  size_t size = 0;
+
+  // Non-constexpr due to reinterpret_cast.
+  template <typename CharT>
+  StringRef(const CharT* data_, size_t size_)
+      : data(reinterpret_cast<const char*>(data_)), size(size_) {
+    // Sanity check for overflowed values.
+    assert(size < 0x8000000000000000ULL);
+  }
+
+  constexpr StringRef(const char* data_, size_t size_) : data(data_), size(size_) {}
+
+  StringRef(const std::string& s) : data(s.data()), size(s.size()) {}  /// NOLINT
+  constexpr explicit StringRef(std::string_view s) : data(s.data()), size(s.size()) {}
+  constexpr StringRef(const char* data_)
+      : StringRef(std::string_view{data_}) {}  /// NOLINT
+  constexpr StringRef() = default;
+
+  bool empty() const { return size == 0; }
+
+  std::string toString() const { return std::string(data, size); }
+  explicit operator std::string() const { return toString(); }
+
+  std::string_view toView() const { return std::string_view(data, size); }
+  constexpr explicit operator std::string_view() const {
+    return std::string_view(data, size);
+  }
+};
+
+using StringRefs = std::vector<StringRef>;
+
+/** Compare strings for equality.
+ * The approach is controversial and does not win in all cases.
+ * For more information, see hash_map_string_2.cpp
+ */
+inline bool compareSSE2(const char* p1, const char* p2) {
+  return 0xFFFF == _mm_movemask_epi8(_mm_cmpeq_epi8(
+                       _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1)),
+                       _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2))));
+}
+
+inline bool compareSSE2x4(const char* p1, const char* p2) {
+  return 0xFFFF ==
+         _mm_movemask_epi8(_mm_and_si128(
+             _mm_and_si128(
+                 _mm_cmpeq_epi8(_mm_loadu_si128(reinterpret_cast<const __m128i*>(p1)),
+                                _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2))),
+                 _mm_cmpeq_epi8(
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1) + 1),
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2) + 1))),
+             _mm_and_si128(
+                 _mm_cmpeq_epi8(
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1) + 2),
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2) + 2)),
+                 _mm_cmpeq_epi8(
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p1) + 3),
+                     _mm_loadu_si128(reinterpret_cast<const __m128i*>(p2) + 3)))));
+}

```

Comment:
does this function has implemented in xsimd?

Reply:
am not clear about xsimd, but I think all `SSE` instructions can be replaced by avx2 except a small part which is not supported or deprecated.